File: code/386-Lexicographical Numbers/386-Lexicographical Numbers.py
# -*- coding: utf-8 -*
"""
LeetCode 386 - Lexicographical Numbers

解题步骤（自动整理）
1. 用 DFS/回溯枚举所有可能选择，并维护当前路径/状态
2. 在递归过程中进行剪枝（如边界、重复、提前失败条件）
3. 到达终止条件时收集结果或返回，并回溯撤销选择
"""
r"""
@author: Kai Chen
@file: 386-Lexicographical Numbers.py
@time: 2020/11/2 9:38
@desc:
Given an integer n, return 1 - n in lexicographical order.

For example, given 13, return: [1,10,11,12,13,2,3,4,5,6,7,8,9].

Please optimize your algorithm to use less time and space.
The input size may be as large as 5,000,000.
"""


class Solution(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        res = []

        def dfs(cur):
            if cur > n:
                return
            res.append(cur)
            for i in range(10):
                dfs(cur * 10 + i)

        for i in range(1, 10):
            dfs(i)
        return res


if __name__ == '__main__':
    s = Solution()
    print(s.lexicalOrder(13))

File: code/386-Lexicographical Numbers/386-Lexicographical Numbers.cpp
/*
LeetCode 386 - Lexicographical Numbers

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/

// Given an integer n, return 1 - n in lexicographical order.

// For example, given 13, return: [1,10,11,12,13,2,3,4,5,6,7,8,9].

// Please optimize your algorithm to use less time and space.
// The input size may be as large as 5,000,000.


class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> res(n, 0);
        int num = 1;
        for (int i = 0; i < n; i ++) {
            res[i] = num;
            if (num * 10 <= n) {
                num *= 10;
            } else {
                while (num % 10 == 9 || num + 1 > n) {
                    num /= 10;
                }
                num ++;
            }
        }
        return res;
    }
};
